### PR TITLE
New dynamic tiling attempt

### PR DESCRIPTION
**Pull request by applmak**
Introduce a mechanism to automatically split up drive images into smaller pieces. Note: this only works for images from drive when a certain config option is set. More general such mechanisms will come later. Bonus: you should only set this for images that are intended to cover the whole wall. There are a lot of caveats and TODOs. Use with caution.

### CODE REVIEW

**Comment by applmak** [comment]
Closes #224

File: demo_modules/slideshow/interfaces.js [hunk 1]
```diff
@@ -50,16 +80,15 @@ export class ClientLoadStrategy {
...
   }
-  newContent(content) {
-    // A notification from the load strategy that new content has been
-    // discovered. The parameter is an array of content identifiers.
+  clipRectForMetadata(metadata, client) {
+    // Returns a clipping rect for the content specified by the metadata for
+    // the specified client. The returned Rectangle should be in the unit-space
+    // of the content (0,0) contentWidth x contentHeight. The strategy can
+    // instead return null to denote that no clipping should occur.
+    return null;
```
Comment (jmegq) [inline demo_modules/slideshow/interfaces.js]: Why not just return the metadata extents in this case? Boo to null, I say.

File: demo_modules/slideshow/slideshow_server.js [hunk 2]
```diff
@@ -24,68 +24,118 @@ limitations under the License.
...
+      debug('Selected', ret, 'for client', client.offset);
+      this.content.push(ret);
+      // Ask the loading strategy to nab some metadata for this content item.
+      const metadata = await this.loadStrategy.metadataForContent(ret, this.contentCache);
+      if (metadata) {
+        debug(`Metadata: ${metadata.width} x ${metadata.height}`);
+        // If it chooses to response with some, we then ask the display
+        // strategy if it wants to clip this content.
+        const clippingRect = this.displayStrategy.clipRectForMetadata(metadata, client);
+        if (clippingRect) {
```
Comment (jmegq) [inline demo_modules/slideshow/slideshow_server.js]: This factoring feels like chooseContent is working too hard in the clipping case, but I don't have a better suggestion.

File: demo_modules/slideshow/interfaces.js (hunk 1 above)
Comment (applmak) [reply demo_modules/slideshow/interfaces.js]: Each of these APIs gives a different strategy the option of giving up on any splitting. Here, we allow the loading strategy to say, you know what, you don't really want this metadata.

File: demo_modules/slideshow/slideshow_server.js (hunk 2 above)
Comment (applmak) [reply demo_modules/slideshow/slideshow_server.js]: Yeah.

File: demo_modules/slideshow/slideshow_server.js [hunk 3]
```diff
@@ -24,68 +24,118 @@ limitations under the License.
...
+     * What to do when new content is downloaded.
+     */
+    handleNewContent(content) {
+      this.content.push(...random.shuffle(content));
+    }
+    /**
+     * Asks the load strategy to load some content and also starts a loop to
+     * load the rest of the content.
+     */
+    async startLoadingContent() {
```
[the remark this answers is not in the document]
Comment (applmak) [reply demo_modules/slideshow/slideshow_server.js]: Yes! I've been thinking about how to do this generally, and I wonder if the lack of this has caused some of the OOMs we've seen on the server. I want to fix that sort of thing more generally, though, which is outside the scope of this PR.

**Pull request merged**